File: leetcode-python/Quests/Array_II.py
class Solution:

    #Array II : Q1. Set Mismatch
    def findErrorNums(self, nums: list[int]) -> list[int]:
        n: int = len(nums)

        expected_sum: int = (n * (n + 1)) // 2
        actual_sum: int = sum(nums)

        duplicate: int = actual_sum - sum(set(nums))
        missing: int = expected_sum + duplicate - actual_sum

        return [duplicate, missing]


    #Array II : Q2. How Many Numbers Are Smaller Than the Current Number
    def smallerNumbersThanCurrent(self, nums: list[int]) -> list[int]:
        sorted_nums: list[int] = sorted(nums)

        return [sorted_nums.index(i) for i in nums]


    #Array II : Q3. Find All Numbers Disappeared in an Array
    def findDisappearedNumbers(self, nums: list[int]) -> list[int]:
        output:list[int] = []
        numbers:set[int] = set(nums)

        for i in range(1, len(nums) + 1):
            if i not in numbers:
                output.append(i)

        return output

File: leetcode-java/src/quest/ArrayII.java
package quest;

import java.util.*;

public class ArrayII
{
    // Array II : Q1. Set Mismatch
    public int[] findErrorNums(int[] nums) {

        int n = nums.length;
        int[] numbers = new int[n + 1];
        int[] result = new int[2];

        for (int i : nums) {
            numbers[i]++;
        }

        for (int i = 1; i <= n; i++) {
            if (numbers[i] == 1) {
                continue;
            }

            if (numbers[i] == 0) {
                result[1] = i;
            }

            if (numbers[i] == 2) {
                result[0] = i;
            }
        }

        return result;

//        // -------------------------------------------------------------------
//        int n = nums.length;
//        int expectedSum = (n * (n + 1)) / 2;
//        int setSum = 0;
//        int arraySum = 0;
//        Set<Integer> set = new HashSet<>();
//
//        for (int i : nums) {
//            arraySum += i;
//            set.add(i);
//        }
//
//        for (int i : set) {
//            setSum += i;
//        }
//
//        return new int[] {arraySum - setSum, expectedSum - setSum};
//        // -------------------------------------------------------------------
    }

    // Array II : Q2. How Many Numbers Are Smaller Than the Current Number
    public int[] smallerNumbersThanCurrent(int[] nums) {

        //Brute force
        int[] result = new int[nums.length];
        int currentNumber = 0;

        for (int i = 0; i < nums.length; i++) {
            currentNumber = nums[i];

            for (int num : nums)
            {
                if (num < currentNumber)
                {
                    result[i]++;
                }
            }
        }

        return result;
    }
}
